From b29a7efc358395ec765d41e1e5a284363f25ec98 Mon Sep 17 00:00:00 2001
From: Alex Johnston <acjohnston@google.com>
Date: Thu, 6 Jan 2022 01:27:36 +0000
Subject: [PATCH] Add caller check to com.android.credentials.RESET

* Only the Car Settings app can reset credentials
  via com.android.credentials.RESET.
* com.android.credentials.INSTALL should still be
  callable by CertInstaller.

Manual testing steps:
* Install certificate via Settings
* Verify unable to reset certificates via test app
  provided in the bug (app-debug.apk)
* Verify able to reset certificates via Settings
* Verify com.android.credentials.INSTALL isn't changed

Bug: 200164168
Test: manual
Change-Id: Ibf6a02a35beed08c769fda9b3044d2e7933186dd
(cherry picked from commit 571f4c5b88c1e1ba3d7f04687d906cc89bfa6dc3)
Merged-In:Ibf6a02a35beed08c769fda9b3044d2e7933186dd
---
 .../security/CredentialStorageActivity.java       | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/src/com/android/car/settings/security/CredentialStorageActivity.java b/src/com/android/car/settings/security/CredentialStorageActivity.java
index 7b8558320..eabfd4046 100644
--- a/src/com/android/car/settings/security/CredentialStorageActivity.java
+++ b/src/com/android/car/settings/security/CredentialStorageActivity.java
@@ -78,7 +78,7 @@ public class CredentialStorageActivity extends FragmentActivity {
 
         Intent intent = getIntent();
         String action = intent.getAction();
-        if (ACTION_RESET.equals(action)) {
+        if (ACTION_RESET.equals(action) && checkCallerIsSelf()) {
             showResetConfirmationDialog();
         } else if (ACTION_INSTALL.equals(action) && checkCallerIsCertInstallerOrSelfInProfile()) {
             Bundle installBundle = intent.getExtras();
@@ -120,6 +120,19 @@ public class CredentialStorageActivity extends FragmentActivity {
         }
     }
 
+    /**
+     * Check that the caller is Settings.
+     */
+    private boolean checkCallerIsSelf() {
+        try {
+            return Process.myUid() == android.app.ActivityManager.getService()
+                    .getLaunchedFromUid(getActivityToken());
+        } catch (RemoteException re) {
+            // Error talking to ActivityManager, just give up
+            return false;
+        }
+    }
+
     /**
      * Check that the caller is either CertInstaller or Settings running in a profile of this user.
      */
-- 
2.33.1

